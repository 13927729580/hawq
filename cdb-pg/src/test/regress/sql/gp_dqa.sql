drop table if exists dqa_t1;
drop table if exists dqa_t2;

create table dqa_t1 (d int, i int, c char, dt date);
create table dqa_t2 (d int, i int, c char, dt date);

insert into dqa_t1 select i%23, i%12, (i%10) || '', '2009-06-10'::date + ( (i%34) || ' days')::interval
from generate_series(0, 99) i;
insert into dqa_t2 select i%34, i%45, (i%10) || '', '2009-06-10'::date + ( (i%56) || ' days')::interval
from generate_series(0, 99) i;

set gp_eager_agg_distinct_pruning=on;
set enable_groupagg=off;

-- Distinct keys are distribution keys
select count(distinct d) from dqa_t1;
select count(distinct d) from dqa_t1 group by i;

select count(distinct d), count(distinct dt) from dqa_t1;
select count(distinct d), count(distinct c), count(distinct dt) from dqa_t1;

select count(distinct d), count(distinct dt) from dqa_t1 group by c;
select count(distinct d), count(distinct dt) from dqa_t1 group by d;

select count(distinct dqa_t1.d) from dqa_t1, dqa_t2 where dqa_t1.d = dqa_t2.d;
select count(distinct dqa_t1.d) from dqa_t1, dqa_t2 where dqa_t1.d = dqa_t2.d group by dqa_t2.dt;

-- Distinct keys are not distribution keys
select count(distinct c) from dqa_t1;
select count(distinct c) from dqa_t1 group by dt;
select count(distinct c) from dqa_t1 group by d;

select count(distinct c), count(distinct dt) from dqa_t1;
select count(distinct c), count(distinct dt), i from dqa_t1 group by i;
select count(distinct i), count(distinct c), d from dqa_t1 group by d;

select count(distinct dqa_t1.dt) from dqa_t1, dqa_t2 where dqa_t1.c = dqa_t2.c;
select count(distinct dqa_t1.dt) from dqa_t1, dqa_t2 where dqa_t1.c = dqa_t2.c group by dqa_t2.dt;
