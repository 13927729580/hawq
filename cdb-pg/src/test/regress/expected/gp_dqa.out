drop table if exists dqa_t1;
drop table if exists dqa_t2;
create table dqa_t1 (d int, i int, c char, dt date);
NOTICE:  Table doesn't have 'DISTRIBUTED BY' clause -- Using column named 'd' as the Greenplum Database data distribution key for this table.
HINT:  The 'DISTRIBUTED BY' clause determines the distribution of data. Make sure column(s) chosen are the optimal data distribution key to minimize skew.
create table dqa_t2 (d int, i int, c char, dt date);
NOTICE:  Table doesn't have 'DISTRIBUTED BY' clause -- Using column named 'd' as the Greenplum Database data distribution key for this table.
HINT:  The 'DISTRIBUTED BY' clause determines the distribution of data. Make sure column(s) chosen are the optimal data distribution key to minimize skew.
insert into dqa_t1 select i%23, i%12, (i%10) || '', '2009-06-10'::date + ( (i%34) || ' days')::interval
from generate_series(0, 99) i;
insert into dqa_t2 select i%34, i%45, (i%10) || '', '2009-06-10'::date + ( (i%56) || ' days')::interval
from generate_series(0, 99) i;
set gp_eager_agg_distinct_pruning=on;
set enable_groupagg=off;
-- Distinct keys are distribution keys
select count(distinct d) from dqa_t1;
 count 
-------
    23
(1 row)

select count(distinct d) from dqa_t1 group by i;
 count 
-------
     9
     8
     8
     9
     8
     8
     9
     8
     8
     9
     8
     8
(12 rows)

select count(distinct d), count(distinct dt) from dqa_t1;
 count | count 
-------+-------
    23 |    34
(1 row)

select count(distinct d), count(distinct c), count(distinct dt) from dqa_t1;
 count | count | count 
-------+-------+-------
    23 |    10 |    34
(1 row)

select count(distinct d), count(distinct dt) from dqa_t1 group by c;
 count | count 
-------+-------
    10 |    10
    10 |    10
    10 |    10
    10 |    10
    10 |    10
    10 |    10
    10 |    10
    10 |    10
    10 |    10
    10 |    10
(10 rows)

select count(distinct d), count(distinct dt) from dqa_t1 group by d;
 count | count 
-------+-------
     1 |     5
     1 |     5
     1 |     4
     1 |     4
     1 |     4
     1 |     5
     1 |     5
     1 |     4
     1 |     4
     1 |     4
     1 |     4
     1 |     5
     1 |     5
     1 |     4
     1 |     4
     1 |     4
     1 |     4
     1 |     5
     1 |     5
     1 |     4
     1 |     4
     1 |     4
     1 |     4
(23 rows)

select count(distinct dqa_t1.d) from dqa_t1, dqa_t2 where dqa_t1.d = dqa_t2.d;
 count 
-------
    23
(1 row)

select count(distinct dqa_t1.d) from dqa_t1, dqa_t2 where dqa_t1.d = dqa_t2.d group by dqa_t2.dt;
 count 
-------
     1
     1
     1
     2
     2
     1
     1
     1
     1
     1
     1
     1
     1
     1
     1
     1
     1
     2
     2
     2
     1
     1
     1
     1
     1
     1
     1
     1
     2
     1
     1
     2
     2
     2
     1
     1
     1
     1
     1
     1
     1
     1
     1
     1
     1
     2
     2
     2
     1
     1
     2
     1
     1
     1
     1
     1
(56 rows)

-- Distinct keys are not distribution keys
select count(distinct c) from dqa_t1;
 count 
-------
    10
(1 row)

select count(distinct c) from dqa_t1 group by dt;
 count 
-------
     3
     3
     3
     3
     3
     3
     3
     3
     2
     3
     3
     3
     3
     3
     3
     3
     3
     3
     3
     3
     3
     3
     3
     3
     3
     3
     3
     3
     3
     3
     3
     3
     3
     2
(34 rows)

select count(distinct c) from dqa_t1 group by d;
 count 
-------
     5
     5
     4
     4
     4
     5
     5
     4
     4
     4
     4
     5
     5
     4
     4
     4
     4
     5
     5
     4
     4
     4
     4
(23 rows)

select count(distinct c), count(distinct dt) from dqa_t1;
 count | count 
-------+-------
    10 |    34
(1 row)

select count(distinct c), count(distinct dt), i from dqa_t1 group by i;
 count | count | i  
-------+-------+----
     5 |     9 |  3
     5 |     8 |  7
     5 |     8 | 11
     5 |     9 |  1
     5 |     8 |  5
     5 |     8 |  9
     5 |     9 |  2
     5 |     8 |  6
     5 |     8 | 10
     5 |     9 |  0
     5 |     8 |  4
     5 |     8 |  8
(12 rows)

select count(distinct i), count(distinct c), d from dqa_t1 group by d;
 count | count | d  
-------+-------+----
     5 |     5 |  3
     5 |     5 |  7
     4 |     4 | 11
     4 |     4 | 15
     4 |     4 | 19
     5 |     5 |  1
     5 |     5 |  5
     4 |     4 |  9
     4 |     4 | 13
     4 |     4 | 17
     4 |     4 | 21
     5 |     5 |  2
     5 |     5 |  6
     4 |     4 | 10
     4 |     4 | 14
     4 |     4 | 18
     4 |     4 | 22
     5 |     5 |  0
     5 |     5 |  4
     4 |     4 |  8
     4 |     4 | 12
     4 |     4 | 16
     4 |     4 | 20
(23 rows)

select count(distinct dqa_t1.dt) from dqa_t1, dqa_t2 where dqa_t1.c = dqa_t2.c;
 count 
-------
    34
(1 row)

select count(distinct dqa_t1.dt) from dqa_t1, dqa_t2 where dqa_t1.c = dqa_t2.c group by dqa_t2.dt;
 count 
-------
    14
    13
    13
    14
    13
    14
    13
    13
    14
    13
    14
    10
    10
    10
    14
    13
    14
    13
    13
    14
    13
    14
    13
    13
    14
    10
    10
    10
    14
    13
    14
    13
    13
    14
    13
    14
    13
    13
    14
    10
    10
    10
    14
    13
    13
    14
    13
    14
    13
    13
    14
    13
    14
    10
    10
    10
(56 rows)

